src/ScheduledUpgradeable.sol

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {IScheduledUpgradeable, ISignatureScheme} from "./interfaces/IScheduledUpgradeable.sol";
+import {ErrorsLib} from "./libraries/ErrorsLib.sol";
+import {BLS} from "bls-solidity/BLS.sol";
+
+/// @title ScheduledUpgradeable
+/// @author Randamu
+/// @notice Abstract contract for scheduling, cancelling, and executing contract upgrades.
+/// @dev Handles BLS (BN254) signature verification for scheduling and cancelling upgrades.
+abstract contract ScheduledUpgradeable is IScheduledUpgradeable, Initializable, UUPSUpgradeable {
+    /// @notice Unique nonce for each message to prevent replay attacks
+    uint256 public currentNonce;
+
+    /// @notice Address of the scheduled implementation upgrade
+    address public scheduledImplementation;
+
+    /// @notice Calldata for the scheduled implementation upgrade
+    bytes internal scheduledImplementationCalldata;
+
+    /// @notice Timestamp for the scheduled implementation upgrade
+    uint256 public scheduledTimestampForUpgrade;
+
+    /// @notice Minimum delay for upgrade operations
+    uint256 public minimumContractUpgradeDelay;
+
+    /// @notice BLS validator used for validating admin threshold signatures for stopping timed upgrades
+    ISignatureScheme public contractUpgradeBlsValidator;
+
+    // ---------------------- Initializer ----------------------
+
+    /// @notice Initializes upgrade scheduling logic.
+    /// @param _contractUpgradeBlsValidator Address of the BLS validator contract
+    /// @param _minimumContractUpgradeDelay Minimum delay for upgrades in seconds
+    function __ScheduledUpgradeable_init(address _contractUpgradeBlsValidator, uint256 _minimumContractUpgradeDelay)
+        internal
+        onlyInitializing
+    {
+        require(_contractUpgradeBlsValidator != address(0), ErrorsLib.ZeroAddress());
+        require(_minimumContractUpgradeDelay >= 2 days, ErrorsLib.UpgradeDelayTooShort());
```

Comment:
There is a somewhat popular point of view according to which exit windows (which upgrade delays are a special case of) should be at least 7 days to be credible. Up to us if we want to abide by this at launch though. https://deficollective.org/blog/introducing-defiscan/

Reply:
Interesting docs! Thanks for sharing @bbjubjub2494 
We have the ability to increase it to 7 days post deployment using the function `setMinimumContractUpgradeDelay(uint256 _minimumContractUpgradeDelay)`.